### PR TITLE
Get parent of embedded object.

### LINKED ISSUE #2742: [Bug]: `EmbeddedObject.Parent` returning `null`
### What happened?

`EmbeddedObject.Parent` is not returning anything after being added to realm

### Repro steps

1. Create a `new RealmObject()` with an `EmbeddedObject`, `IList<EmbeddedObject`, or `IDictionary<TKey, EmbeddedObject>` property
2. Add `RealmObject` to Realm
3. Populate the aforementioned property with a `new EmbeddedObject()`
4. Accessing the `Parent` property of the added `EmbeddedObject` returns null

### Version

10.7.1

### What SDK flavour are you using?

Local Database only

### What type of application is this?

Console/Server

### Client OS and version

Windows 10 Home 20H2 19042.1348

### Code snippets

Here's a small program for reproducing the bug

```cs
public class Test_EmbeddedObjects {

    class Item : RealmObject {
        public string? Label { get; set; }
        public ChildItem? ChildItem { get; set; }
        public IList<ChildItem> ChildItemList { get; } = null!;
        public IDictionary<string, ChildItem> ChildItemDict { get; } = null!;
    }

    class ChildItem : EmbeddedObject {
        public string? Label { get; set; }
    }

    public static void Main() {
        var config = new InMemoryConfiguration( "meemoo.db" ) {
            Schema = new Type[] { typeof( Item ), typeof( ChildItem ) }
        };
        var realm = Realm.GetInstance( config );

        var item = new Item { Label = "A" };
        item.ChildItem = new ChildItem { Label = "b" };
        realm.Write( () => {
            realm.Add( item );
            item.ChildItemList.Add( new ChildItem { Label = "c" } );
            item.ChildItemDict.Add( "d", new ChildItem { Label = "d" } );
        } );

        Console.WriteLine( item.ChildItem.Parent != null); // False
        Console.WriteLine( item.ChildItemList.First().Parent != null); // False
        Console.WriteLine( item.ChildItemDict.Values.First().Parent != null); // False
    }
}
```

### Stacktrace of the exception/crash you're getting

_No response_

### Relevant log output

_No response_

### PR DESCRIPTION
**Pull request by elle-j**
## Description

Fixes #2742.

Makes `Parent` accessible on `IEmbeddedObject`s and returns `null` for unmanaged embedded objects.

##  TODO

* [x] Changelog entry
* [x] Tests (if applicable)

### CODE REVIEW

File: Realm/Realm/DatabaseTypes/Accessors/IRealmAccessor.cs [hunk 1]
```diff
@@ -141,6 +141,13 @@ public interface IRealmAccessor
         IQueryable<T> GetBacklinks<T>(string propertyName)
             where T : IRealmObjectBase;
 
+        /// <summary>
+        /// Gets the parent of the <see cref="IEmbeddedObject">embedded object</see>. It can be either another
+        /// <see cref="IEmbeddedObject">embedded object</see> or a standalone <see cref="IRealmObject">realm object</see>.
```
Comment (nirinchev) [inline Realm/Realm/DatabaseTypes/Accessors/IRealmAccessor.cs:146]:
```suggestion
        /// <see cref="IEmbeddedObject">embedded object</see>, a standalone <see cref="IRealmObject">realm object</see>, or an <see cref="IAsymmetricObject>asymmetric object</see>.
```

File: Realm/Realm/DatabaseTypes/Accessors/UnmanagedAccessor.cs [hunk 2]
```diff
@@ -93,6 +93,11 @@ public void SetValueUnique(string propertyName, RealmValue val)
             throw new NotImplementedException("This should not be used for now");
         }
 
+        public IRealmObjectBase GetParent()
+        {
+            return null;
+        }
```
Comment (nirinchev) [inline Realm/Realm/DatabaseTypes/Accessors/UnmanagedAccessor.cs:99]: Slightly shorter

```suggestion
        public IRealmObjectBase GetParent() => null;
```

File: Realm/Realm/DatabaseTypes/IRealmObjectBase.cs [hunk 3]
```diff
@@ -96,6 +96,11 @@ public interface IAsymmetricObject : IRealmObjectBase
     /// </summary>
     public interface IEmbeddedObject : IRealmObjectBase
     {
+        /// <summary>
+        /// Gets the parent of the <see cref="IEmbeddedObject">embedded object</see>. It can be either another
+        /// <see cref="IEmbeddedObject">embedded object</see> or a standalone <see cref="IRealmObject">realm object</see>.
```
Comment (nirinchev) [inline Realm/Realm/DatabaseTypes/IRealmObjectBase.cs:101]:
```suggestion
        /// <see cref="IEmbeddedObject">embedded object</see>, a standalone <see cref="IRealmObject">realm object</see>, or an <see cref="IAsymmetricObject">asymmetric object</see>.
```

**Review by nirinchev — APPROVED** [review]
Good stuff. Only a few nits from me.

**Comment by elle-j** [comment]
Tests for having an embedded object as a `RealmValue` and accessing its parent will be added separately as Core currently throws a `LogicError` for illegal type.

E.g. consider the test:
```csharp
[Test]
public void EmbeddedObject_WhenParentAccessedInRealmValue_ReturnsParent()
{
    var parent = new AllTypesObject
    {
        RequiredStringProperty = "parent",
        RealmValueProperty = new EmbeddedIntPropertyObject()
    };

    _realm.Write(() =>
    {
        _realm.Add(parent);
    });

    Assert.That(parent, Is.EqualTo(parent.RealmValueProperty.AsRealmObject<EmbeddedIntPropertyObject>().Parent));
}
```
Then Core's `obj.cpp` will throw because `type` is `ColumnType::Type::Mixed` rather than`ColumnType::Type::Link`:

```cpp
Obj Obj::create_and_set_linked_object(ColKey col_key, bool is_default)
{
    // ...

    ColumnType type = col_key.get_type();            // Mixed
    if (type != col_type_Link)
        throw LogicError(LogicError::illegal_type);

    // ...
}
```

**Comment by nirinchev** [comment]
Hm... interesting - it seems like `RealmValue` should not be able to contain an embedded object. This looks like an oversight on our end and we should probably guard better against it - i.e. throwing a more meaningful error message.

File: Realm/Realm/Schema/ObjectSchema.cs [hunk 4]
```diff
@@ -80,7 +80,7 @@ public enum ObjectType : byte
         /// Gets a value indicating whether this <see cref="ObjectSchema"/> describes an embedded object.
         /// </summary>
         /// <value><c>true</c> if the schema pertains to an <see cref="EmbeddedObject"/> instance; <c>false</c> otherwise.</value>
-        [Obsolete("Check against RealmSchemaType instead.")]
+        [Obsolete("Check against BaseType instead.")]
```
Comment (LaPeste) [inline Realm/Realm/Schema/ObjectSchema.cs:83]: Ops, a leftover from the asymmetric objects. Thank you for fixing it.

**Review by LaPeste — APPROVED** [review]

**Comment by nirinchev** [comment]
I can see there are several test failures that are legitimate issues with the tests. Let's take a look at those tomorrow - let me know if you need a crash course on interpreting the CI test results.

**Comment by elle-j** [comment]
Yeah it's the asymmetric objects with sync, I'll fix those now. Was using a type from Realm.Tests/Database/TestObjects.cs.

**Pull request merged**